docs/design/operator.md

```diff
@@ -0,0 +1,407 @@
+created in different namespace that are targetting same csi driver name.
+
+```yaml
+---
+kind: CephCSIDriver
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: "<prefix>.<driver_type>.csi.ceph.com"
+  namespace: operator-namespace
+spec:
+  deploymentName: csi-cephfsplugin-provisioner
+  daemonsetName: csi-cephfsplugin
+  Spec:
+    fsGroupPolicy: File
+    encryption:
+      configMapRef:
+        name: encryption-config-map-name
+    plugin:
+      priorityClassName: system-node-critical
+      updateStrategy:
+        type: RollingUpdate
+        rollingUpdate:
+          maxUnavailable: 1
+      labels:
+        app: cephfs-plugin
+      annotations:
+        k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+    provisioner:
+      labels:
+        app: ceph-fs-provisioner
+      annotations:
+        k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+      provisionerReplicas: 2
+      leaderElection:
+        leaseDuration: 100
+        renewDeadline: 100
+        retryPeriod: 10
+      attachRequired: true
+    liveness:
+      metricsPort: 8000
+    deployCSIAddons: false
+  cephFsKernelClient: true
+  cephFsKernelMountOptions: ms_mode=secure
+status:
+  phase: Failed
+  message: Failed to create cephfs csi driver
+  reason: csi driver with same name already exists in the cluster
+```
+
+### CephCSICephConfig CRD
+
+Stores connection and configuration details for a single single Ceph cluster.
+Provides common connection and configuration information for multiple CSI
+drivers
+
+```yaml
+---
+kind: CephCSICephCluster
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: ceph-cluster-1
+  namespace: operator-namespace
+spec:
+  monitors:
+  - 10.98.44.171:6789
+  - 10.98.44.172:6789
+  - 10.98.44.173:6789
+  ReadAffinity:
+    crushLocationLabels:
+    - kubernetes.io/hostname
+    - topology.kubernetes.io/region
+    - topology.kubernetes.io/zone
+  cephFS:
+    kernelMountOptions: readdir_max_bytes=1048576,norbytes
+    fuseMountOptions: debug
+  rbd:
+    mirrorDaemonCount: 2
+  cephConfig: |-
+    [global]
+    auth_cluster_required = none
+    auth_service_required = none
+    auth_client_required = none
+    rbd_validate_pool = false
+status: {}
+```
+
+### CephCSIConfig CRD
+
+Contains details about CephFS subvolume group or RADOS namespace and a
+reference to CephCSIConfig to link storage configurations to Ceph cluster.
+
+```yaml
+---
+kind: CephCSIConfig
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: storage
+  namespace: operator-namespace
+spec:
+  cephClusterRef:
+    name: ceph-cluster-1
+  subvolumeGroup: csi
+  radosNamespace: rados-test
+status:
+  phase: Succeeded
+  message: successfully linked to CephClusterRef
+```
+
+### CephCSIConfigRecovery
+
+The CephCSIConfigRecovery CR contains the local and the remote cephcluster
+configuration which will help cephcsi to identify the blockpool alias to
+consume the volume from when doing CSI operation.
+
+```yaml
+---
+kind: CephCSIConfigRecovery
```

Comment:
Is this a new concept for the csi driver? Or how is it specified in the driver today? I remember discussions about this feature, but don't recall if it's already there.

Reply:
No this is an existing configmap Rook is created today for the RBDPoolMapping. https://github.com/rook/rook/blob/8dc88c39b105c79f8f5678f83dd23414411daa40/pkg/operator/ceph/csi/peermap/config.go#L51-L54